File: techstore-backend/warehouse-service/src/main/java/com/techstore/warehouse/service/SupplierService.java
package com.techstore.warehouse.service;

import java.util.List;

import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.techstore.warehouse.constant.WarehouseStatus;
import com.techstore.warehouse.dto.request.SupplierCreateRequest;
import com.techstore.warehouse.dto.request.SupplierUpdateRequest;
import com.techstore.warehouse.dto.response.SupplierResponse;
import com.techstore.warehouse.entity.Supplier;
import com.techstore.warehouse.exception.AppException;
import com.techstore.warehouse.exception.ErrorCode;
import com.techstore.warehouse.mapper.SupplierMapper;
import com.techstore.warehouse.repository.SupplierRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Slf4j
public class SupplierService {

    private final SupplierRepository supplierRepo;
    private final SupplierMapper supplierMapper;

    @PreAuthorize("hasAnyRole('ADMIN','STAFF')")
    @Transactional
    public SupplierResponse create(SupplierCreateRequest req) {
        log.info("Creating supplier with phone: {}", req.getPhone());

        if (supplierRepo.findByPhone(req.getPhone()).isPresent()) {
            throw new AppException(ErrorCode.SUPPLIER_EXISTED);
        }

        Supplier supplier = supplierMapper.toEntity(req);
        supplier.setStatus(WarehouseStatus.ACTIVE.name());

        return supplierMapper.toResponse(supplierRepo.save(supplier));
    }

    @PreAuthorize("hasAnyRole('ADMIN','STAFF')")
    @Transactional
    public SupplierResponse update(Long id, SupplierUpdateRequest req) {
        log.info("Updating supplier with id: {}", id);

        Supplier supplier = supplierRepo.findById(id).orElseThrow(() -> new AppException(ErrorCode.SUPPLIER_NOT_FOUND));

        if (req.getPhone() != null && !req.getPhone().equals(supplier.getPhone())) {
            supplierRepo.findByPhone(req.getPhone()).ifPresent(s -> {
                throw new AppException(ErrorCode.SUPPLIER_EXISTED);
            });
        }

        supplierMapper.updateEntityFromRequest(req, supplier);
        return supplierMapper.toResponse(supplierRepo.save(supplier));
    }

    @PreAuthorize("hasAnyRole('ADMIN','STAFF')")
    public SupplierResponse getById(Long id) {
        return supplierMapper.toResponse(
                supplierRepo.findById(id).orElseThrow(() -> new AppException(ErrorCode.SUPPLIER_NOT_FOUND)));
    }

    @PreAuthorize("hasAnyRole('ADMIN','STAFF')")
    public SupplierResponse getByPhone(String phone) {
        return supplierMapper.toResponse(
                supplierRepo.findByPhone(phone).orElseThrow(() -> new AppException(ErrorCode.SUPPLIER_NOT_FOUND)));
    }

    @PreAuthorize("hasAnyRole('ADMIN','STAFF')")
    public List<SupplierResponse> getAll() {
        return supplierRepo.findAll().stream().map(supplierMapper::toResponse).toList();
    }

    @PreAuthorize("hasAnyRole('ADMIN','STAFF')")
    public List<SupplierResponse> getByStatus(String status) {
        return supplierRepo.findByStatus(status).stream()
                .map(supplierMapper::toResponse)
                .toList();
    }

    @PreAuthorize("hasAnyRole('ADMIN','STAFF')")
    public List<SupplierResponse> searchByName(String name) {
        return supplierRepo.findByNameContainingIgnoreCase(name).stream()
                .map(supplierMapper::toResponse)
                .toList();
    }

    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public void delete(Long id) {
        log.info("Deleting supplier with id: {}", id);

        Supplier supplier = supplierRepo.findById(id).orElseThrow(() -> new AppException(ErrorCode.SUPPLIER_NOT_FOUND));

        supplierRepo.delete(supplier);
    }

    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public SupplierResponse updateStatus(Long id, String status) {
        log.info("Updating supplier status for id: {} to status: {}", id, status);

        Supplier supplier = supplierRepo.findById(id).orElseThrow(() -> new AppException(ErrorCode.SUPPLIER_NOT_FOUND));

        supplier.setStatus(status);
        return supplierMapper.toResponse(supplierRepo.save(supplier));
    }
}

File: techstore-frontend/techstore-admin/src/app/features/supplier/supplier.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../../shared/models/api-response.model';
import { SupplierResponse } from './models/supplier.model';
import { SupplierCreateRequest, SupplierUpdateRequest } from './models/supplier-request.model';

@Injectable({ providedIn: 'root' })
export class SupplierService {
  private baseUrl = environment.warehouseUrl + '/suppliers';

  constructor(private http: HttpClient) {}

  create(req: SupplierCreateRequest):
    Observable<ApiResponse<SupplierResponse>> {
    return this.http.post<ApiResponse<SupplierResponse>>(this.baseUrl, req);
  }

  update(id: number, req: SupplierUpdateRequest):
    Observable<ApiResponse<SupplierResponse>> {
    return this.http.put<ApiResponse<SupplierResponse>>(`${this.baseUrl}/${id}`, req);
  }

  getById(id: number):
    Observable<ApiResponse<SupplierResponse>> {
    return this.http.get<ApiResponse<SupplierResponse>>(`${this.baseUrl}/${id}`);
  }

  getByPhone(phone: string):
    Observable<ApiResponse<SupplierResponse>> {
    return this.http.get<ApiResponse<SupplierResponse>>(`${this.baseUrl}/phone/${phone}`);
  }

  getAll():
    Observable<ApiResponse<SupplierResponse[]>> {
    return this.http.get<ApiResponse<SupplierResponse[]>>(this.baseUrl);
  }

  getByStatus(status: string):
    Observable<ApiResponse<SupplierResponse[]>> {
    return this.http.get<ApiResponse<SupplierResponse[]>>(`${this.baseUrl}/status/${status}`);
  }

  searchByName(name: string):
    Observable<ApiResponse<SupplierResponse[]>> {
    const params = new HttpParams().set('name', name);
    return this.http.get<ApiResponse<SupplierResponse[]>>(`${this.baseUrl}/search`, { params });
  }

  delete(id: number):
    Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.baseUrl}/${id}`);
  }

  updateStatus(id: number, status: string):
    Observable<ApiResponse<SupplierResponse>> {
    const params = new HttpParams().set('status', status);
    return this.http.put<ApiResponse<SupplierResponse>>(`${this.baseUrl}/${id}/status`, null, { params });
  }
}